### PR TITLE
make stop_ atomic and check it in GatherMinibatch()

### PR DESCRIPTION
**Pull request by borg323**
Also remove mutex from IsSearchActive() and GUARDED_BY annotation from stop_.

### CODE REVIEW

File: src/mcts/search.cc [hunk 1]
```diff
@@ -519,7 +519,6 @@ void Search::RunBlocking(size_t threads) {
 }
 
 bool Search::IsSearchActive() const {
-  Mutex::Lock lock(counters_mutex_);
   return !stop_;
```
Comment (mooskagh) [inline src/mcts/search.cc]: `!stop_.load()`
or `!stop_.load(std::memory_order_acquire)` to make it even faster, at all places where `stop_.load()` is called (no performance penalty due to being atomic at all on x86)

Similarly, all places where value is stored into the atomic, should be `stop_.store(value)`, or `stop_.store(value, std::memory_order_release)`.

File: src/mcts/search.cc [hunk 2]
```diff
@@ -692,6 +688,8 @@ void SearchWorker::GatherMinibatch() {
       --minibatch_size;
       ++number_out_of_order;
     }
+    // Check for stop at the end so we have at least one node.
+    if (search_->stop_) return;
```
Comment (mooskagh) [inline src/mcts/search.cc]: `if (search_->stop_.load())`

File: src/mcts/search.cc [hunk 3]
```diff
@@ -692,6 +689,8 @@ void SearchWorker::GatherMinibatch() {
       --minibatch_size;
       ++number_out_of_order;
     }
+    // Check for stop at the end so we have at least one node.
+    if (search_->stop_.load(std::memory_order_acquire)) return;
```
Comment (Tilps) [inline src/mcts/search.cc]: I think this check should also be added above the continue for processing a collision.

**Comment by Tilps** [comment]
I wonder if this may not be very effective due to prefetching then going and filling the remaining slots in the early out batch.

**Comment by mooskagh** [comment]
Indeed, we also should not prefetch anything if `stop_`, and maybe stop prefetching in the middle too.

File: src/mcts/search.cc [hunk 4]
```diff
@@ -542,17 +541,17 @@ void Search::WatchdogThread() {
...
+      watchdog_cv_.wait_for(lock.get_raw(), remaining_time, [this]() {
+        return stop_.load(std::memory_order_acquire);
+      });
     }
     MaybeTriggerStop();
   }
   MaybeTriggerStop();
 }
 
 void Search::FireStopInternal() REQUIRES(counters_mutex_) {
```
Comment (mooskagh) [inline src/mcts/search.cc]: Doesn't really require `counters_mutex_` anymore, so this annotation can be removed.

Comment (borg323) [reply src/mcts/search.cc]: Done

**Pull request merged**